bob-common/src/mapper.rs

```diff
@@ -121,28 +119,98 @@ impl Virtual {
+        nodes: &[Node],
+        target_nodes: &[Node],
+        offset: usize,
+    ) -> SupportIndexResult {
+        let mut avail = 0;
+        for i in 0..nodes.len() {
+            let node = &nodes[i];
+            if node.connection_available() && target_nodes.iter().all(|n| n.index() != node.index())
+            {
+                if avail == offset {
+                    return SupportIndexResult {
+                        index: Some(i),
+                        avail: avail + 1,
+                    };
+                }
+                avail += 1;
+            }
+        }
+        return SupportIndexResult { index: None, avail };
+    }
+
+    /// Look for an offset respecting the uniform distribution
+    fn find_support_node_offset(nodes: &[Node], target_nodes: &[Node], offset: usize) -> usize {
+        if target_nodes.len() >= nodes.len() {
+            return 0;
+        }
+        let mut res = Self::try_find_support_node_offset_at_one_pass(nodes, target_nodes, offset);
+        if res.index == None && res.avail > 0 {
+            let mut curr_offset = offset % res.avail;
+            while res.index == None && res.avail > 0 {
+                res = Self::try_find_support_node_offset_at_one_pass(
+                    nodes,
+                    target_nodes,
+                    curr_offset,
+                );
+                curr_offset = if curr_offset > res.avail {
+                    curr_offset - res.avail
+                } else {
+                    0
+                }
+            }
+        }
+        match res.index {
+            Some(index) => index,
+            None => offset % nodes.len(),
+        }
+    }
+
     pub fn get_support_nodes(&self, key: BobKey, count: usize) -> Vec<&Node> {
         debug_assert!(count <= self.nodes.len());
         if count == 0 {
             return vec![];
         }
         trace!("get target nodes for given key");
         let target_nodes = self.get_target_nodes_for_key(key);
+        debug_assert!(target_nodes.iter().map(|n| n.index()).collect::<HashSet<NodeId>>().len() == target_nodes.len());
+        if target_nodes.len() >= self.nodes.len() {
+            return vec![];
+        }
         trace!("extract indexes of target nodes");
         trace!("nodes available: {}", self.nodes.len());
-        let offset = self.support_nodes_offset.fetch_add(1, Ordering::Relaxed);
-        let mut support_nodes = Vec::with_capacity(count);
-        for (id, node) in self.nodes.iter().skip(offset % self.nodes.len()) {
-            if target_nodes.iter().all(|i| i.index() != *id) {
+        let mut support_nodes: Vec<&Node> = Vec::with_capacity(count);
+
+        // This offset search preserves the uniform distribution of aliens
+        let starting_index = Self::find_support_node_offset(
+            &self.nodes,
+            target_nodes,
+            self.support_nodes_offset.fetch_add(1, Ordering::Relaxed)
+                % (self.nodes.len() - target_nodes.len()),
```

Comment:
You should pass the offset without calculating the remainder, since at this level it is necessary to abstract from the offset search logic implemented in `find_support_node_offset`